# 1.2 Objects

- Python is an object-oriented language and **classes** from the basis for all data types.
- Python has built-in classes, such as the `int` class for integers.
- The `float` class for floating-point values,
- and the `str` class for character strings.

## 1.2.1 Identifiers, Objects and the Assignment Statement

- The most important of all Python commands is an **assignment statement**, such as -

In [1]:
temperature = 98.6

- This command establishes temperature as an identifier (also known as name), and then associates it with the **object** expressed on the right-hand side of the equal sign, in this case a floating-point object with value 98.6.

|float|
|---|
|98.6|

⬆️
temperature

**Figure 1.1:** The identifier `temperature` references an instance of the float class having value 98.6.

### 1. Identifier

- Identifiers in Python are case sensitive.
- so temperature and Temperature are distinct names.
- Identifier can be composed of almost any combination of letters, numerals, and underscore characters (or more general Unicode characters).
- The primary restriction are that an identifier cannot begin with a numeral (thus 9lives is an illegal name).
- There are 33 specially reserved words that cannot be used as identifiers.

**Table 1.1:** A list of the reserved words in Python. These names cannot be used as identifier.

|Reserved Words 7|Reserved Words 7|Reserved Words 7|Reserved Words 7|Reserved Words 5|
|---|---|---|---|---|
|None|for|pass|class|try|
|and|while|break|assert|except|
|or|if|continue|def|finally|
|not|elif|from|del|raise|
|in|else|import|lambda|with|
|is|True|global|yield||
|as|False|nonlocal|return||

- The semantics of a Python identifier is most similar to a **reference variable** in Java or a **pointer variable** in C++.
- Each identifier is implicitly associated with the **memory address** of the object to which it refers.
- A Python identifier may be assigned to a special object named `None`, serving a similar purpose to a null reference in Java or C++.

### 2. Dynamically Typed

- Unlike Java and C++, Python is a **dynamically typed** language, as there is no advance declaration associating an identifier with a particular data type.
- Dynamic typing means that the type of the variable is determined only during runtime.
- Due to dynamic typing, in Python the same variable can have a different type at different times during the execution. Dynamic typing allows for flexibility in programming, but with a price in performance.
- An identifier can be associated with any type of object, and it can later be reassigned to another object of the same (or different) type.
- Although an identifier has no declared type, the object to which it refers has a definite type.
- In our first example, the character 98.6 are recognized as a floating-point literal, and thus the identifier temperature is associated with an instance of the float class having that value.

### 3. Literal

- Python literals are quantities/ notations whose value does not change during the execution of a program.
- Literals in Python are nothing but a succinct way of representing the data types. In simpler words, it is the way to represent a fixed value in our source code. They can either be numbers, text, boolean or any other form of data.
- We have the concept of **constants** in Python as well! They are known as literals in Python.
- There are five types of literal in Python, which are as follows-
  1. **String Literals :** Single-line String, Multi-line String.
  2. **Numeric Literals :** Integer, Float, Complex, Long.
  3. **Boolean Literals :** True, False.
  4. **Special Literals :** None
  5. **Literal Collections :** List Literals, Tuple Literals, Dictionary Literals, Set Literals.

### 4. Strongly Typed

- Strong typing means that variables do have a type and that the type matters when performing operations on a variable.
- Due to strong typing, types need to be compatible with respect to the operand when performing operations. For example Python allows one to add an integer and a floating point number, but adding an integer to a string produces error.

### 5. Alias

- A programmer can establish an **alias** by assigning a second identifier to an existing object.

**Example :** Identifier temperature and original are aliases for the same object -

In [2]:
temperature = 98.6
original = temperature

**Figure 1.2:** Identifiers temperature and original are aliases for the same object.

|float|
|---|
|98.6|
⬆️ ⬆️ original
temperature

- Once an alias has been established, either name can be used to access the underlying object.
- If that object supports behaviors that affect its state, changes enacted through one alias will be apparent when using the other alias (because they refer to the same object).
- However, if one of the names is reassigned to a new value using a subsequent assignment statement, that does not affect the aliased object, rather it breaks the alias.

In [3]:
temperature = 98.6
original = temperature
temperature = temperature + 5.0
print(temperature, original)

103.6 98.6


- The execution of this command begins with the evaluation of the expression on the right-hand side of the = operator.
- That expression, `temperature + 5.0`, is evaluated based on the existing binding of the name temperature, and so the result has value 103.6. That result is stored as a new floating-point instance, and only then is the name on the left-hand side of the assignment statement, temperature, (re)assigned to the result.
- Of particular note, this last command had no effect on the value of the existing float instance that identifier original continues to reference.

**Figure 1.3:** The temperature identifier has been assigned to a new value, while original continues to refer to the previously existing value.

|float|
|---|
|103.6|
⬆️ temperature

|float|
|---|
|98.6|
⬆️ original

- Here, is the example of the deep copy and shallow copy

In [4]:
temperature = 98.6
original = temperature
original = 10
print(temperature, original)

98.6 10


In [5]:
d1 = {'a':1, 'b':2}
d2 = d1
print(id(d1)==id(d2))
del d2['b']     # del d1['b']
print(d1, d2)

True
{'a': 1} {'a': 1}


- Here, original and d2 both second alias refers to same object which assigned but `original` assigned to new object and d2 is still refers to same object so any changes in d2 reflects changes in both (d1 and d2).

## 1.2.2 Creating and Using Objects

### 1. Instantiation

- The process of creating a new instance of a class is known as **instantiation**.
- In general, the syntax for instantiating an object is to invoke the **constructor** of a class.
- For Example, If there were a class named `Widget`, we could create an instance of that class using a syntax such as `w = Widget()`, assuming that the constructor does not require any parameters. If the constructor does require parameters, we might use a syntax such as `Widget(a, b, c)` to construct a new instance.

**Literals :**

In [6]:
temperature = 98.6

- Literal form for designating new instances.
- For example, the command temperature = 98.6 results in the creation of a new instance of the **float** class.
- The term 98.6 in that expression is a literal form.

- From Programmer's perspective, yet another way to indirectly create a new instance of a class is to call a function that creates and returns such as instance.
- For example, Python has a built-in function named sorted() that takes a sequence of comparable elements as a parameter and returns a new instance of the **list** class containing those elements in sorted order.

### 2. Calling Method

- Python supports traditional functions that are invoked with a syntax such as sorted(data), in which case data is a parameter sent to the function.
- Python's classes may also define one or more **methods** (also known as **member functions**), which are invoked on a specific instance of a class using dot (".") operator.
- For example, Python's list class has a method named sort that can be invoked with a syntax such as data.sort(). This particular method rearranges the content of the list so that they are sorted.

In [8]:
data = list()
data.sort()

- The expression to the left of the dot identifies the object upon which the method is invoked.
- Often, this will be an identifier (e.g. data), but we can use the dot operator to invoke a method upon the immediate result of some other operation.
- For example, if response identifies a string instance -

In [9]:
response = "Younger Brother"
response.lower().startswith('y')

True

- The syntax response.lower().startswith('y') first evaluates the method call, response.lower(), which itself returns a new string instance, and then the startswith('y') method is called on that intermediate string.

- When using a method of a class, it is important to understand its behavior.

### 3. Accessor

- Some methods return information about the state of an object, but do not change that state. These are known as **accessors**.

### 4. Mutators / Update Methods

- Other methods, such as the sort method of the list class, do change the state of an object. these methods are known as **mutators** or **update method**.

## 1.2.3 Python's Built-In Classes

### Immutable

- A class is **immutable** if each object of that class has a fixed value upon instantiation that cannot subsequently be changed.
- For example - The float class is immutable.
- Once an instance has been created, its value cannot be changed (although an identifier referencing that object can be reassigned to a different value).

**Table 1.2:** Commonly used built-in classes for Python

|Class|Description|Immutable?|
|---|---|---|
|**bool**|Boolean value|✔️|
|**int**|integer (arbitrary magnitude)|✔️|
|**float**|floating-point number|✔️|
|**list**|mutable sequence of objects||
|**tuple**|immutable sequence of objects|✔️|
|**str**|character string|✔️|
|**set**|unordered set of distinct objects||
|**frozenset**|immutable from of set class|✔️|
|**dict**|associative mapping (aka dictionary||

- Literal forms (such as 98.6) exist for most of the built-in classes, and all of the classes support a traditional constructor from that creates instances that are based upon one or more existing values.

### 1. bool Class

- The **bool** class is used to manipulate logical (Boolean) values, and the only two instances of that class are expressed as the literals **True** and **False**.
- The default constructor, `bool()`, return False, but there is no reason to use that syntax rather than more direct literal form.

In [10]:
flag = bool()
print(flag)

False


- Python allows the creation of a Boolean value from a non boolean type using the syntax `bool(foo)` for value foo.

In [1]:
foo = 12      # non-zero value
print(bool(foo))

True


**The interpretation depends on the type of the parameter :**

- Number evaluate to False if zero, and True if nonzero.

In [3]:
non_zero = 0     # zero
print(bool(non_zero))

num = 23      # Number
print(bool(num))

False
True


- Sequences and other container types, such as strings and lists, evaluate to False if empty and True if nonempty.

In [4]:
sequenceBool1 = bool(list())       # Empty list sequence datatype
print(sequenceBool1)

sequenceBool2 = bool('Hello World!')    # Empty string sequence
print(sequenceBool2)

False
True


#### Application

- An important application of this interpretation is the use of a non boolean value as a condition in a control structure.
- **Example :**

In [5]:
done = False
while not done:
    print("1")
    done = True

1


### 2. int Class

- The `int` and `float` classes are the primary numeric types in Python.
- The `int` class is designed to represent integer values with arbitrary magnitude.
- Unlike Java and C++, which different precisions (i.e. int, short, long), Python automatically chooses the internal representation for an integer based upon the magnitude of its value.
- Typical literals for integers include 0, 137 and -23.
- In some contexts, it is convenient to express an integer value using binary, octal and hexadecimal
  - That can be done by using a prefix of the number 0 and then a character to describe the base.
  - **Example :** such as literals are -
  - `0b10111` -> binary
  - `0o52` -> octal
  - `0x7f` -> hexadecimal
- The integer constructor, `int()`, returns value 0 by default.
- But this constructor can be used to construct an integer value based upon an existing value of another type.
  - **Example :**
  - if `f` represents a floating-point value, the syntax `int(f)` produces the truncated value of `f`.

In [6]:
var1 = int(3.14)    # produces value 3
var2 = int(3.99)    # produces value 3
var3 = int(-3.9)    # produces value -3

print(var1, var2, var3)

3 3 -3


- The constructor can also be used to parse a string that is presumed to represent an integer value (such as one entered by a user).
- **Example :**
- It 's' represents a string then `int(s)` produces the integral value that string represents.

In [7]:
var = int('137')
print(var)

137


- If an invalid string is given as a parameter as in `int('Hello')`, a ValueError is raised.
- By default string must use base 10.
- If conversion from a different base is desired, that base can be indicated as a second, optional, parameter.
- **Example :**

In [9]:
v = int('7f', 16)       # evaluates "7f" as hexadecimal to the integer 127

print(v)

127


### 3. float Class

- The `float` class is the floating-point type in Python, using a fixed-precision representation.
- Its precision is more akin to  a double in Java or C++, rather than those language float type.
- Floating-point equivalent of an integral number can be expressed as 2.0 or 2. , the trailing zero is optional.
- One other from of literal for floating-point values uses scientific notation.
- **Example :** literal 6.022e23 or 6.022 x 10^23
- The constructor form of `float()` has default return 0.0

In [10]:
fl = float()
print(fl)

0.0


- When given a parameter, the constructor attempts to return the equivalent floating-point value.
- **Example :** `float(2)` returns value `2.0`

In [11]:
f2 = float(2)
print(f2)

2.0


- If the parameter to the constructor is a string, as with `float(3.14)` gives value 3.14
- float('hello'), It attempts to parse that string as a floating-point value, raising a ValueError as an exception.

## Sequence Types: The list, tuple, and str Classes

- The list, tuple and str classes are sequence type in Python, representing a collection of values in which the order is significant.

### 4. list Class

- A list instance stores a sequence of objects.
- A list is a referential structure, as it technically stores a sequence of references to its elements.
- Elements of a list many be arbitrary objects (including the None object).
- list are array-based sequences and are zero-indexed, thus length of list n has elements indexed from 0 to (n-1) inclusive.
- Python uses the character [] as delimiters for a list literal, with [] itself being an empty list.
- **Example :**

In [12]:
v2 = ['red', 'green', 'blue']   # is a list containing three string instances.
print(v2)

['red', 'green', 'blue']


- The contents of a list literal need not be expressed as literals.
- If identifiers a and b have been established, then syntax [a, b] is legitimate.
- The list constructor produces an empty list by default `list()`
- However, the constructor will accept any parameter that is of an iterable type.
  - **Example :** Standard-Iterable type (string, list, tuple, set, dictionaries)
  - `list('hello')` produces a list of individual characters ['h', 'r', 'l', 'l', 'o'] because an existing list is itself iterable
  - The syntax - `backup = list(data)` can be used to construct a new list instance referencing the same contents as the original.

|addresses :|.|.|.|.|.|.|.|.|.|.|.|
|---|---|---|---|---|---|---|---|---|---|---|---|
|primes (objects):|2|3|5|7|11|13|17|19|23|29|31|

### 5. tuple Class

- The tuple classes provides an immutable version of a sequence. amd therefore its instances have an internal representation that may be more streamlined than that of a list.
- parenthesis `()` delimit a tuple, with `()` being an empty tuple.
- There is one important subtlety -
  - To express a tuple of length one as a literal, a comma must be placed after the element, but within the parenthesis.
    - **Example :** `(17, )` is a one-element tuple is that without trailing comma, the expression (17) is viewed as a simple parenthesized numeric expression.

### 6. str Class

- Python's set class represents the mathematical notation of a set, namely a collection. of elements, without duplicates, and without an inherent order to these elements.
- The major advantage of using a set, as opposed to a list, is that, it has a highly optimized method for checking whether a specific element is contained in the set.
- This is based on a data structure known as a hash table.
- However, There are two important restrictions due to the algorithmic underpinnings.
  1. The set does
  2. 

### 7. The set and frozenset Classes

- G

### 8. dict Class

- H